### ISSUE TITLE
Longhorns floods k8s api server with 40k incorrect replica CRs, resulting in unhealthy 00-core-connectivity-k8s k8s cluster

### ISSUE THREAD

**Issue by gberche-orange** [orange-cloudfoundry/paas-templates#2070]
### Expected behavior

* As a paas-templates operator
* In order to ensure make use of core connectivity K8s cluster
* I need longhorn to be robust to failures

### Observed behavior
 
Following the two failure modes,
* #2060 
* #2062 due to tcp checksum offload bug

We observe:
* high load on the k8s api database (pxc-mysql instance), high memory usage, high number of mysql connections
* k3s server process memory usage grows from 5MB up to  30 GB
* regular ooms on the k3s-server node resulting in
    * k3s server and k8s workloads to be killed
    * bosh agent to be killed
       * the bosh instance group appearing as unresponsive
* slow sql query warnings in 
   * k3s-server kine logs
   * pxc-mysql 
* k8s api response timeouts failing k8s workloads including controllers (such as longhorn, gitlab...)
* k8s api database has 204k longhorn replica entries
   * created after recovery from  #2060 
   * created with a high rate: 
   * 6k to 21k replicas per pvc

<details>

![htop of server/0 when k3s server memory is growing](https://github.com/orange-cloudfoundry/paas-templates/assets/4748380/ead6fe43-8bff-48d6-8ba8-1c0887b8c49d)

Sample ooms kills

```
2023-09-13T07:10:43.946188+00:00 17606ee1-6544-4936-b258-d7eaaee210b4 kernel: [406468.634715] oom-kill:constraint=CONSTRAINT_NONE,nodemask=(null),cpuset=6a8ea06d9594cf33a34187381ca15d7e9aadcbc7f8cc5d864386a56715dc9a64,mems_allowed=0,global_oom,task_memcg=/kubepods/besteffort/pod7d59e6ad-9268-41ae-841b-df757e9f9a9f/207b32b7d75d8ae14a98a0fca78e24a6bc94b6db0a6faa513f1549a151fcc1a5,task=csi-snapshotter,pid=585417,uid=0
2023-09-13T07:10:43.946192+00:00 17606ee1-6544-4936-b258-d7eaaee210b4 kernel: [406468.634759] Out of memory: Killed process 585417 (csi-snapshotter) total-vm:741100kB, anon-rss:6444kB, file-rss:0kB, shmem-rss:0kB, UID:0 pgtables:172kB oom_score_adj:1000
2023-09-13T07:10:43.946193+00:00 17606ee1-6544-4936-b258-d7eaaee210b4 kernel: [406468.770933] coredns invoked oom-killer: gfp_mask=0x140cca(GFP_HIGHUSER_MOVABLE|__GFP_COMP), order=0, oom_score_adj=998

j:0
2023-09-13T07:10:49.820531+00:00 17606ee1-6544-4936-b258-d7eaaee210b4 kernel: [406477.992751] Out of memory: Killed process 575275 (runsvdir) total-vm:2808kB, anon-rss:100kB, file-rss:0kB, shmem-rss:0kB, UID:0 pgtables:52kB oom_score_adj:0
2023-09-13T07:10:49.820781+00:00 17606ee1-6544-4936-b258-d7eaaee210b4 kernel: [406478.008373] Out of memory: Killed process 575733 (sleep) total-vm:5768kB, anon-rss:88kB, file-rss:0kB, shmem-rss:0kB, UID:1000 pgtables:52kB oom_score_adj:0
2023-09-13T07:10:49.821143+00:00 17606ee1-6544-4936-b258-d7eaaee210b4 kernel: [406478.010381] Out of memory: Killed process 575720 (tini) total-vm:2784kB, anon-rss:92kB, file-rss:0kB, shmem-rss:0kB, UID:1000 pgtables:48kB oom_score_adj:0
2023-09-13T07:10:49.821397+00:00 17606ee1-6544-4936-b258-d7eaaee210b4 kernel: [406478.015052] Out of memory: Killed process 575271 (agetty) total-vm:6172kB, anon-rss:88kB, file-rss:0kB, shmem-rss:0kB, UID:0 pgtables:48kB oom_score_adj:0
2023-09-13T07:10:49.821620+00:00 17606ee1-6544-4936-b258-d7eaaee210b4 kernel: [406478.048943] Out of memory: Killed process 575288 (svlogd) total-vm:2804kB, anon-rss:84kB, file-rss:0kB, shmem-rss:0kB, UID:0 pgtables:44kB oom_score_adj:0
2023-09-13T07:10:49.821791+00:00 17606ee1-6544-4936-b258-d7eaaee210b4 kernel: [406478.056896] Out of memory: Killed process 575286 (runsv) total-vm:2656kB, anon-rss:88kB, file-rss:0kB, shmem-rss:0kB, UID:0 pgtables:40kB oom_score_adj:0
2023-09-13T07:10:49.821791+00:00 17606ee1-6544-4936-b258-d7eaaee210b4 kernel: [406478.063752] Out of memory: Killed process 575285 (runsv) total-vm:2656kB, anon-rss:88kB, file-rss:0kB, shmem-rss:0kB, UID:0 pgtables:40kB oom_score_adj:0
2023-09-13T07:10:49.821792+00:00 17606ee1-6544-4936-b258-d7eaaee210b4 kernel: [406478.176427] Out of memory: Killed process 586056 (run) total-vm:1800kB, anon-rss:28kB, file-rss:0kB, shmem-rss:0kB, UID:0 pgtables:40kB oom_score_adj:0
2023-09-13T07:10:49.821799+00:00 17606ee1-6544-4936-b258-d7eaaee210b4 kernel: [406478.197120] Out of memory: Killed process 584021 (k3s-server) total-vm:32622368kB, anon-rss:31345812kB, file-rss:0kB, shmem-rss:0kB, UID:0 pgtables:61824kB oom_score_adj:-999
```

Sample slow query logs

```
server/11aa3763-7a95-47ad-8089-7399a97a2d0d:~# cd /var/vcap/sys/log/k3s-server
server/11aa3763-7a95-47ad-8089-7399a97a2d0d:/var/vcap/sys/log/k3s-server# tail -f k3s-server.stderr.log

time="2023-09-13T11:37:04Z" level=info msg="Slow SQL (started: 2023-09-13 11:36:34.901221551 +0000 UTC m=+531.476785757) (total time: 29.591448932s):  SELECT * FROM ( SELECT ( SELECT MAX(rkv.id) AS id FROM kine AS rkv), ( SELECT MAX(crkv.prev_revision) AS prev_revision FROM kine AS crkv WHERE crkv.name = 'compact_rev_key'), kv.id AS theid, kv.name, kv.created, kv.deleted, kv.create_revision, kv.prev_revision, kv.lease, kv.value, kv.old_value FROM kine AS kv JOIN ( SELECT MAX(mkv.id) AS id FROM kine AS mkv WHERE mkv.name LIKE ? AND mkv.id <= ? AND mkv.id > ( SELECT MAX(ikv.id) AS id FROM kine AS ikv WHERE ikv.name = ? AND ikv.id <= ?) GROUP BY mkv.name) AS maxkv ON maxkv.id = kv.id WHERE kv.deleted = 0 OR ? ) AS lkv ORDER BY lkv.theid ASC LIMIT 501 : [[/registry/longhorn.io/replicas/% 587159684 /registry/longhorn.io/replicas/02-longhorn/pvc-231ccc4e-310a-465d-9d26-908e2e0fa78e-r-b9c65a34 587159684 false]]"
```

```
mysql/6ae4b76e-20bf-4a22-8935-02df27c5be5d:/var/vcap/sys/log/pxc-mysql# less mysql_slow_query.log 

# Time: 2023-09-13T14:25:48.344555Z
# User@Host: k3s-db-user[k3s-db-user] @  [192.168.116.224]  Id: 1333199
# Schema: k3sdb  Last_errno: 0  Killed: 0
# Query_time: 11.929811  Lock_time: 0.000081  Rows_sent: 501  Rows_examined: 1303323  Rows_affected: 0  Bytes_sent: 2664812
SET timestamp=1694615148;
SELECT *
                FROM (
                        SELECT (
                SELECT MAX(rkv.id) AS id
                FROM kine AS rkv), (
                SELECT MAX(crkv.prev_revision) AS prev_revision
                FROM kine AS crkv
                WHERE crkv.name = 'compact_rev_key'), kv.id AS theid, kv.name, kv.created, kv.deleted, kv.create_revision, kv.prev_revision, kv.lease, kv.value, kv.old_value
                        FROM kine AS kv
                        JOIN (
                                SELECT MAX(mkv.id) AS id
                                FROM kine AS mkv
                                WHERE
                                        mkv.name LIKE '/registry/longhorn.io/replicas/%'
                                        
                AND
                mkv.id <= 587278510 AND
                mkv.id > (
                        SELECT MAX(ikv.id) AS id
                        FROM kine AS ikv
                        WHERE
                                ikv.name = '/registry/longhorn.io/replicas/02-longhorn/pvc-231ccc4e-310a-465d-9d26-908e2e0fa78e-r-2909aa90' AND
                                ikv.id <= 587278510)
                                GROUP BY mkv.name) AS maxkv
                                ON maxkv.id = kv.id
                        WHERE
                                kv.deleted = 0 OR
                                0
                ) AS lkv
                ORDER BY lkv.theid ASC
                 LIMIT 501;
```

most slow queries are related to longhorn
```
mysql/6ae4b76e-20bf-4a22-8935-02df27c5be5d:/var/vcap/sys/log/pxc-mysql# grep mkv.name mysql_slow_query.log | sort | uniq -c | sort -n
      1                                         mkv.name LIKE '/registry/apiextensions.k8s.io/customresourcedefinitions/%'
      2                                         mkv.name LIKE '/registry/pods/%'
     24                                         mkv.name LIKE '/registry/secrets/%'
    793                                         mkv.name LIKE '/%'
   1565                                         mkv.name LIKE '/registry/events/%'
  15698                                         mkv.name LIKE '/registry/longhorn.io/replicas/%'

ysql/6ae4b76e-20bf-4a22-8935-02df27c5be5d:/var/vcap/sys/log/pxc-mysql# grep ikv.name mysql_slow_query.log | sort | uniq -c | sort -nr
   1432                                 ikv.name = '/registry/longhorn.io/replicas/02-longhorn/pvc-e4630a8c-8f0b-4852-a9f4-4e1998a41339-r-ee47d18f' AND
   1000                                 ikv.name = '/registry/longhorn.io/replicas/02-longhorn/pvc-231ccc4e-310a-465d-9d26-908e2e0fa78e-r-8baa2614' AND
    829                                 ikv.name = '/registry/longhorn.io/replicas/02-longhorn/pvc-231ccc4e-310a-465d-9d26-908e2e0fa78e-r-2649d527' AND
    723                                 ikv.name = '/registry/longhorn.io/replicas/02-longhorn/pvc-231ccc4e-310a-465d-9d26-908e2e0fa78e-r-89da51eb' AND
    642                                 ikv.name = '/registry/longhorn.io/replicas/02-longhorn/pvc-231ccc4e-310a-465d-9d26-908e2e0fa78e-r-5e90d5fd' AND
    576                                 ikv.name = '/registry/longhorn.io/replicas/02-longhorn/pvc-e4630a8c-8f0b-4852-a9f4-4e1998a41339-r-57e2d094' AND


mysql/6ae4b76e-20bf-4a22-8935-02df27c5be5d:/var/vcap/sys/log/pxc-mysql# grep ikv.name mysql_slow_query.log | grep -i longhorn | wc -l
13437
mysql/6ae4b76e-20bf-4a22-8935-02df27c5be5d:/var/vcap/sys/log/pxc-mysql# grep ikv.name mysql_slow_query.log | grep -i -v longhorn | wc -l
1054
```

Query times and data returned for slow queries are high enough to freeze a mysql thread/process for minutes

```
mysql/6ae4b76e-20bf-4a22-8935-02df27c5be5d:/var/vcap/sys/log/pxc-mysql# grep 'Query_time' mysql_slow_query.log

# Query_time: 31.953721  Lock_time: 0.000101  Rows_sent: 501  Rows_examined: 1349547  Rows_affected: 0  Bytes_sent: 2664868
# Query_time: 30.181370  Lock_time: 0.000073  Rows_sent: 501  Rows_examined: 1336806  Rows_affected: 0  Bytes_sent: 2664880
# Query_time: 32.149475  Lock_time: 0.000100  Rows_sent: 501  Rows_examined: 1336747  Rows_affected: 0  Bytes_sent: 2664880
# Query_time: 32.258104  Lock_time: 0.000154  Rows_sent: 501  Rows_examined: 1352795  Rows_affected: 0  Bytes_sent: 2664772
# Query_time: 32.355318  Lock_time: 0.000102  Rows_sent: 501  Rows_examined: 1352739  Rows_affected: 0  Bytes_sent: 2664772
# Query_time: 21.964996  Lock_time: 0.000065  Rows_sent: 501  Rows_examined: 812733  Rows_affected: 0  Bytes_sent: 2664844
# Query_time: 1812.682079  Lock_time: 0.000064  Rows_sent: 3  Rows_examined: 1286325  Rows_affected: 0  Bytes_sent: 0
# Query_time: 2007.033941  Lock_time: 0.000091  Rows_sent: 3  Rows_examined: 1325757  Rows_affected: 0  Bytes_sent: 0
# Query_time: 1755.972087  Lock_time: 0.000079  Rows_sent: 4  Rows_examined: 1271809  Rows_affected: 0  Bytes_sent: 0
```` 

Running the modified statement with a count show that more than 173k longhorn replicas in the database

```
mysql> SELECT count(*)          FROM (                  SELECT (                SELECT MAX(rkv.id) AS id                FROM kine AS rkv), (            SELECT MAX(crkv.prev_revision) AS prev_revision                 FROM kine AS crkv               WHERE crkv.name = 'compact_rev_key'), kv.id AS theid, kv.name, kv.created, kv.deleted, kv.create_revision, kv.prev_revision, kv.lease, kv.value, kv.old_value                        FROM kine AS kv                         JOIN (                          SELECT MAX(mkv.id) AS id                                FROM kine AS mkv                                WHERE                                mkv.name LIKE '/registry/longhorn.io/replicas/%'                                                        AND             mkv.id <= 587293602 AND                 mkv.id > (                      SELECT MAX(ikv.id) AS id                        FROM kine AS ikv                        WHERE                           ikv.name = '/registry/longhorn.io/replicas/02-longhorn/pvc-e4630a8c-8f0b-4852-a9f4-4e1998a41339-r-de99d0c3' AND                              ikv.id <= 587293602)                            GROUP BY mkv.name) AS maxkv                             ON maxkv.id = kv.id                     WHERE                           kv.deleted = 0 OR                            0               ) AS lkv                ORDER BY lkv.theid ASC           LIMIT 10;
+----------+
| count(*) |
+----------+
|   173,058 |
+----------+
1 row in set (7.85 sec)
```

Downloaded the etcd db content using mysql and extracted json documents associated to longhorn replicas

```
$ cat  unsorted-all-replicas-1409-0924.txt | sed 's#.*creationTimestamp\":\"##g' | cut -c1-20 > all-replicas-1409-0924-creation-dates.txt
2023-08-14T09:43:49Z
2023-08-14T09:43:51Z
2023-08-14T09:43:52Z
2023-08-14T09:44:09Z
2023-08-14T09:44:11Z
2023-08-14T09:44:12Z
2023-09-05T18:03:41Z
2023-09-05T18:03:41Z
2023-09-05T18:03:41Z


023-09-06T13:02:11Z
2023-09-06T13:02:12Z
2023-09-06T13:02:12Z
2023-09-06T13:02:13Z
```

Sample faulty replica massively created: no diskid, attached node
```json
{
  "apiVersion": "longhorn.io/v1beta2",
  "kind": "Replica",
  "metadata": {
    "creationTimestamp": "2023-09-06T13:02:13Z",
    "generation": 1,
    "labels": {
      "longhorn.io/backing-image": "",
      "longhorndiskuuid": "",
      "longhornnode": ""
    },
    "name": "pvc-b5b6f759-2e3c-4bc1-b6d4-c194760e298f-r-8260f72a",
    "namespace": "02-longhorn",
    "ownerReferences": [
      {
        "apiVersion": "longhorn.io/v1beta2",
        "kind": "Volume",
        "name": "pvc-b5b6f759-2e3c-4bc1-b6d4-c194760e298f",
        "uid": "e3170ceb-b93c-4e71-89fc-3ffddca7ab6e"
      }
    ],
    "uid": "77273538-b09a-4ad4-98d7-d06840326b7a"
  },
  "spec": {
    "active": true,
    "backingImage": "",
    "baseImage": "",
    "dataDirectoryName": "",
    "dataPath": "",
    "desireState": "stopped",
    "diskID": "",
    "diskPath": "",
    "engineImage": "longhornio/longhorn-engine:v1.4.0",
    "engineName": "pvc-b5b6f759-2e3c-4bc1-b6d4-c194760e298f-e-84648df6",
    "failedAt": "",
    "hardNodeAffinity": "",
    "healthyAt": "",
    "logRequested": false,
    "nodeID": "",
    "rebuildRetryCount": 5,
    "revisionCounterDisabled": false,
    "salvageRequested": false,
    "unmapMarkDiskChainRemovedEnabled": false,
    "volumeName": "pvc-b5b6f759-2e3c-4bc1-b6d4-c194760e298f",
    "volumeSize": "21,474,836,480"
  }
}
```

In comparison, sample healthy replica updated just before the longhorn incident started

```json
{
  "apiVersion": "longhorn.io/v1beta2",
  "kind": "Replica",
  "metadata": {
    "creationTimestamp": "2023-08-14T09:44:12Z",
    "deletionGracePeriodSeconds": 0,
    "deletionTimestamp": "2023-09-05T20:35:07Z",
    "finalizers": [
      "longhorn.io"
    ],
    "generation": 5,
    "labels": {
      "longhorn.io/backing-image": "",
      "longhorndiskuuid": "18ffe9ff-aefe-46e3-a75c-57c3f077f081",
      "longhornnode": "minio-agents-0",
      "longhornvolume": "pvc-fdbe21ce-c5cd-40fc-bb72-35b009286ab9"
    },
    "name": "pvc-fdbe21ce-c5cd-40fc-bb72-35b009286ab9-r-21c2e67c",
    "namespace": "02-longhorn",
    "ownerReferences": [
      {
        "apiVersion": "longhorn.io/v1beta2",
        "kind": "Volume",
        "name": "pvc-fdbe21ce-c5cd-40fc-bb72-35b009286ab9",
        "uid": "93cbd8e9-aa52-48ad-8b77-1fa502a95f50"
      }
    ],
    "uid": "c18a8325-6406-438d-b500-7071a9dcc404"
  },
  "spec": {
    "active": true,
    "backingImage": "",
    "baseImage": "",
    "dataDirectoryName": "pvc-fdbe21ce-c5cd-40fc-bb72-35b009286ab9-64783947",
    "dataPath": "",
    "desireState": "stopped",
    "diskID": "18ffe9ff-aefe-46e3-a75c-57c3f077f081",
    "diskPath": "/var/vcap/store/longhorn/",
    "engineImage": "longhornio/longhorn-engine:v1.4.0",
    "engineName": "pvc-fdbe21ce-c5cd-40fc-bb72-35b009286ab9-e-4339f7fc",
    "failedAt": "2023-09-05T20:04:49Z",
    "hardNodeAffinity": "",
    "healthyAt": "2023-08-14T09:44:32Z",
    "logRequested": false,
    "nodeID": "minio-agents-0",
    "rebuildRetryCount": 0,
    "revisionCounterDisabled": false,
    "salvageRequested": false,
    "unmapMarkDiskChainRemovedEnabled": false,
    "volumeName": "pvc-fdbe21ce-c5cd-40fc-bb72-35b009286ab9",
    "volumeSize": "21474836480"
  },
  "status": {
    "conditions": [
      {
        "lastProbeTime": "",
        "lastTransitionTime": "2023-08-14T09:44:12Z",
        "message": "",
        "reason": "",
        "status": "True",
        "type": "InstanceCreation"
      }
    ],
    "currentImage": "",
    "currentState": "stopped",
    "evictionRequested": false,
    "instanceManagerName": "",
    "ip": "",
    "logFetched": false,
    "ownerID": "bastion-agents-2",
    "port": 0,
    "salvageExecuted": false,
    "started": false,
    "storageIP": ""
  }
}
```

</details>

### Workaround

Delete incorrect replica CRs from the K8S api database:
* possibly remove the unnecessary longhorn pvcs (low priority bastion containers)
* identify the id of the correct replicas to preserve
* stop the k3s agetn
* delete the incorrect replicas from the K8S api database
* observe the K8S api recover
* observe the K8S workloads recover
* observe the longhorn

<details>

Still observing 185k events in the k8s api

```
mysql/6ae4b76e-20bf-4a22-8935-02df27c5be5d:~# cat /var/vcap/store/etcd-entries.txt | grep events | wc -l
185196
mysql/6ae4b76e-20bf-4a22-8935-02df27c5be5d:~# cat /var/vcap/store/etcd-entries.txt | grep -v events | wc -l
9209
```

</details>

### Future improvements

* [ ] Reproduce the issue against longhorn 1.4.3
* [ ] Report the issue upstream
* [ ] Add limits to the k8s api against risky CR api groups
* [ ] Improve recovery procedure for `00-core-connectivity-k8s` cluser (to be moved into a distinct issue with `epic` tag)
   * [ ] Up-to-date git clones of all repos in inception vms
   * [ ] Regular snapshots of k8s api databases content as json documents: 
      * [ ] Prepare extraction of K8S resource specs from k8s api inventory snapshots
   * [ ] Recover/Bootstrap procedure from missing `00-core-connectivity-k8s` deployment
      * [ ] Application of key K8S resource specs from k8s api inventory snapshots
      * [ ] Fallback for unhealthy battery-included components
         * [ ] egress internet access
         * [ ] containerd OCI image pulling
         * [ ] flux source repositories supporting kustomizations (currently git repos based on gitlab)
            * [ ] git tea ?
            * [ ] embedded s3 buckets (similar to weaveworks 
   * [ ] Recovery procedure to access key observability components in case of unhealthy `00-core-connectivity-k8s` cluster (prometheus, grafana)
      * [ ] Breakglass http proxy from inception vm
* [ ] Improve observability of k8s clusters (to be moved into a distinct issue with `epic` tag)
   * [ ] Observability of k8s api database:
      * [ ] Slow queries alerts
      * [ ] Database size
      * [ ] Conservative alerting on database capacity
         * [ ] load average
         * [ ] disk space
   * [ ] Observability of k8s inventory (through prometheus ?)
      * [ ] Total number of resources
      * [ ] Numer of resources per categories

### Affected releases

* x.y
* earlier versions

**Comment by gberche-orange** [comment]
Steps to reproduce the problem:
* [x] check existing alerts
* [B] check ethtool prometheus metrics. Blocked by https://github.com/orange-cloudfoundry/paas-templates/issues/2091
* [B] check k8s api server metrics. Blocked by https://github.com/orange-cloudfoundry/paas-templates/issues/2092
* [ ] turn on the tcp-offload
* [ ] monitor k8s api traffic in real time, see workaround in https://github.com/orange-cloudfoundry/paas-templates/issues/2092

**Comment by gberche-orange** [comment]
Possibly better post-v56 deployment with longhorn 1.4.3.

Moving this epic to P1 with 58 milestone for middle-term review.
